public/app/features/variables/editor/selectors.ts

```diff
@@ -0,0 +1,37 @@
+import {
+  AdHocVariableEditorState,
+  DataSourceVariableEditorState,
+  QueryVariableEditorState,
+  VariableEditorState,
+} from './reducer';
+
+export function getAdhocVariableState(
+  editorState: VariableEditorState
+): VariableEditorState<AdHocVariableEditorState> | null {
+  const { extended, ...rest } = editorState;
+
+  if (extended && 'dataSources' in extended) {
+    return {
+      extended,
+      ...rest,
+    };
+  }
+
+  return null;
+}
+
+export function getDatasourceVariableState(editorState: VariableEditorState): DataSourceVariableEditorState | null {
+  if (editorState.extended && 'dataSourceTypes' in editorState.extended) {
+    return editorState.extended;
+  }
+
+  return null;
+}
```

Comment:
Or, a "simpler" selector, more geared to be used in a render function.

Reply:
I prefer this selector over the one used in Adhoc.

Reply:
How do you feel about then how it's actually used in the components? https://github.com/grafana/grafana/pull/44749/files#diff-7c68d8e773d0949597127594a57f6dc632f7194755786fe546fd37f52f6a9eeb

 - AdhocVariableEditor uses a more complicated selector, but allows the [view _using_ it](https://github.com/grafana/grafana/pull/44749/files#diff-79c3921c3bfa84fefe087b545132b25ff88852dbd8b26dd26efb8e6f3743156b) to be simpler and less room for mistakes
   - however this one require a [bunch of generics in the state type](https://github.com/grafana/grafana/pull/44749/files#diff-4a995a538d05b0805440eca25f743a6521027148fec55afcd1ba60e57a97c40aR22-R24) which makes it more complicated
 - DatasourceVariableEditor's selector is simpler, but [requires the guard in more places in the view](https://github.com/grafana/grafana/pull/44749/files#diff-7c68d8e773d0949597127594a57f6dc632f7194755786fe546fd37f52f6a9eeb) and/or more changes to the component
   - could be mitigated by using it in the `mapStateToProps` function with a new mapped prop


thoughts?